### DISCUSSION TITLE
How can I introduce randomness to the schedule?

### DISCUSSION THREAD

**Discussion by Etienne-M** [renovatebot/renovate#21305] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

32.239.0

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

We have fairly high amount of projects (~80), we already limit renovate to 2 PRs per project and to once per month. But still, having 75 PRs open at the same time is not really productive: either we address all of them right away and it kills our productivity, or we address them little by little, but it floods our open PR tracking and we lose track of the PRs opened by our colleagues

Ideally, I would not go through every project and manually assign a different day of the month to each project. I would like something like the `H` syntax that Jenkins offers for cron job, this way I could set something like `0 0 H * *` and have the PRs open on a random day of the month.

Is someone aware of a way to do this?

### What you tried so far.

I looked at the https://www.npmjs.com/package/@breejs/later code, but it does not seem to support the 'H' syntax for cron jobs like Jenkins does

**Comment by HonkingGoose** [comment, marked as answer]
> Ideally, I would not go through every project and manually assign a different day of the month to each project. I would like something like the `H` syntax that Jenkins offers for cron job, this way I could set something like `0 0 H * *` and have the PRs open on a random day of the month.

Even if Renovate or `breejs` supported random times, having Renovate open the PRs at a random day and time would _not_ fix your problem, because you're still getting :

- the same amount of updates
- the updates at a time when you, or your team aren't ready (even worse, now it's _random_)
- a cluttered overview because of the open Renovate PRs (fixable with the right `author:` filter, see below)

I recommend you:

- Update your version of Renovate
- Use the Dependency Dashboard Approval Workflow
- Research further Noise Reduction strategies

## Filtering out PRs from Renovate bot

You can filter out Renovate bot PRs, with the `author:` filter plus `-` as a logical NOT: [^author-github-filter]. Try this filter:

`is:pr is:open -author:renovate`

## Update your version of Renovate

> ### If you're self-hosting Renovate, tell us what version of Renovate you run.
>
> 32.239.0

I recommend you update your version of Renovate to get new features and fixes. 😉 But first, read the major version release notes. [^releasenotes]

## Dependency Dashboard Approval workflow

Have you seen the "Dependency Dashboard Approval" workflow docs? [^dashboard-approval] [^config-option] You can tell Renovate to only give you updates after you select the update from the Dependency Dashboard. In short:

1. Put the the `:dependencyDashboardApproval` preset [^preset] in the config for the repositories you want to "slow down Renovate", this also automatically enables the Dependency Dashboard issue
1. Request updates from Renovate when you're ready for them
1. If you don't want to slow down Renovate for all packages in a repository, use `packageRules` with a suitable `match*` rule to only use the Dependency Dashboard Approval flow for specific packages

## Noise reduction docs

We also have docs about reducing the noise from Renovate updates [^noise], that covers strategies like:

- Package grouping
- Scheduling Renovate (you're already doing this)
- Automerging
- Lock file considerations

[^dashboard-approval]: [Renovate docs, Dependency Dashboard Approval](https://docs.renovatebot.com/getting-started/use-cases/#dependency-dashboard-approval)
[^preset]: [Renovate docs, `:dependencyDashboardApproval` preset](https://docs.renovatebot.com/presets-default/#dependencydashboardapproval)
[^config-option]: [Renovate docs, `dependencyDashboardApproval` config option](https://docs.renovatebot.com/configuration-options/#dependencydashboardapproval)
[^noise]: [Renovate docs, Noise reduction](https://docs.renovatebot.com/noise-reduction/)
[^releasenotes]: [Renovate docs, Release notes for major versions of Renovate](https://docs.renovatebot.com/release-notes-for-major-versions/)
[^author-github-filter]: [GitHub Docs, search by author](https://docs.github.com/en/search-github/searching-on-github/searching-issues-and-pull-requests#search-by-author)

**Reply by Etienne-M** [reply]
First of all, thanks a lot for the detailed answer! This is a lot of useful information, and I really appreciate the time you've put into answering my question 🙂 

If I understand correctly your message, there's no way currently to introduce randomness into the schedule. Fair enough, it's a very niche feature. I'd argue it would be a good feature to add; much like scheduling too many jobs at the same time can overwhelm a computer, opening too many PRs at the same time can overwhelm the humans who have to merge them.

To give more context, our issue is that we have a fairly large amount of projects (>80) to maintain for a relatively small team (6). We use Renovate to give us a much needed nudge to update our dependencies. But the issue isn't that we cannot merge them all. It's they arrive all at the same time. I was hoping to avoid setting them all on a different schedule manually, and instead set `0 0 H * *` for all of them and they would naturally be spread throughout the month.

To address specifically your suggestions:
> Update your version of Renovate

Duly noted 🙂 

> Dependency Dashboard Approval workflow

I like this workflow for some select projects, but in our case it might not work. Given that we use renovate to get the ball rolling, having to go in the project to tick a box to get the PR opened does not really help us much. What we really need is a PR that is opened for us to go: "Ah yes, we have to update some dependencies in this project". Otherwise we will forget about these projects.

> Noise reduction docs

You have good solutions to offer, I'm especially interested in "Automerging"

So to sum up, since the feature does not exists, I will manually set a different schedule on all my projects 😔 But I have some interesting improvements to bring back to my team. Again, thanks a lot for your answer

**Reply by viceice** [reply]
you should run renovate more regularly, so not all PR are created at same time. you can also use commits per run limit, so on a single renovate run you get a limited amount of new commits, which is automatically blocking new PR's

**Reply by HonkingGoose** [reply]
> you can also use commits per run limit, so on a single renovate run you get a limited amount of new commits, which is automatically blocking new PR's

Read the [`prCommitsPerRunLimit` docs](https://docs.renovatebot.com/self-hosted-configuration/#prcommitsperrunlimit) to learn more.
